packages/fiber/src/three-types.ts

```diff
@@ -53,10 +53,18 @@ export type ThreeElement<T extends ConstructorRepresentation> = Mutable<
 >
 
 type ThreeExports = typeof THREE
+
+// Detect conflicts between Three.js and React types.
+type DuplicateKeys<T, U> = Extract<keyof T, keyof U>
+type Conflicts = DuplicateKeys<JSX.IntrinsicElements, { [K in keyof ThreeExports as Uncapitalize<K>]: any }>
+
+// Create a new type that maps Three.js exports to JSX tags, with conflicts prefixed with 'three'.
 type ThreeElementsImpl = {
-  [K in keyof ThreeExports as Uncapitalize<K>]: ThreeExports[K] extends ConstructorRepresentation
-    ? ThreeElement<ThreeExports[K]>
-    : never
+  [K in keyof ThreeExports as K extends string
+    ? Uncapitalize<K> extends Conflicts
+      ? `three${Capitalize<K>}`
+      : Uncapitalize<K>
+    : never]: ThreeExports[K] extends ConstructorRepresentation ? ThreeElement<ThreeExports[K]> : never
 }
```

Comment:
Why the `K extends string`? I think this could be:

```tsx
// Conflicting keys between Three.js and React DOM types (e.g. <line />).
type Conflicts = Extract<keyof JSX.IntrinsicElements, Uncapitalize<keyof ThreeExports>>

// Create a new type that maps Three.js exports to JSX tags, with conflicts prefixed with 'three'.
type ThreeElementsImpl = {
  [K in keyof ThreeExports as Uncapitalize<K> extends Conflicts
    ? `three${K}`
    : Uncapitalize<K>]: ThreeExports[K] extends ConstructorRepresentation ? ThreeElement<ThreeExports[K]> : never
}
```

Reply:
Yes this is fine, I was being extra careful since keys are not necessarily strings but I think in the Three exports case it always is.